GH-40333: [Docs] Improve env var docs for ARROW_USER_SIMD_LEVEL (#40374)
apache/arrow: docs/source/cpp/env_vars.rst

.. envvar:: ARROW_USER_SIMD_LEVEL

   The maximum SIMD optimization level selectable at runtime.  Useful for
   comparing the performance impact of enabling or disabling respective code
   paths or working around situations where instructions are supported but are

>    either buggy or not performant.

Reviewer: Just curious, what does `buggy` here means? ( I'm not familiar with SIMD ). Would that's the case that the arch don't implement the instr well? Or just don't have this instr, do user need disable it?

Author: Thanks for taking a look. There may be a better word here...By buggy I was referring to the comment in https://github.com/apache/arrow/issues/40333#issuecomment-1977142577 mentioning situations where like running under a hypervisor where cpu detection reports that an instruction is supported but it's not handled correctly by the hypervisor. I'm not familiar with this so I don't know what "not handled correctly" really means.

Maybe instead of buggy, say "where instructions are supported but are not performant or cause other issues"?

Reviewer: Looks nice, we can wait for @pitrou and others' idea here

Reviewer 2: I agree with "where instructions are supported but are not performant or cause other issues"?

Author: Thanks! Change made in 5ad7282f7bc3581825098d3789fa0af0c3f0af5f.

   platforms currently do not implement any form of runtime dispatch.

   .. note::
      In addition to runtime-selected SIMD optimizations dispatch, Arrow C++ can

>       also be compiled with SIMD optimizations that are cannot be disabled at

Reviewer 3: ```suggestion
      also be compiled with SIMD optimizations that cannot be disabled at
```

Author: Thank you!

Author: Sorry, got distracted by a phone call. Will push up in a sec.

Author: See 0aab71c1e3f87218571aaca6bcb09d23df37fb87.

   platforms currently do not implement any form of runtime dispatch.

   .. note::
      In addition to runtime dispatch, Arrow's build system supports two
      compile-time CMake variables for controlling SIMD: ``ARROW_SIMD_LEVEL``
      and ``ARROW_RUNTIME_SIMD_LEVEL``.

>       The ``ARROW_SIMD_LEVEL`` variable sets the maximum SIMD level at the
>       compiler level.  Unlike runtime dispatch, compile-time SIMD optimizations

Reviewer 4: For clarity, maybe we should share the mechanism. Also, I think it's more appropriate to call this the *minimum* SIMD level, as it is older computers that won't support this.
```suggestion
      The ``ARROW_SIMD_LEVEL`` variable sets the minimum supported SIMD level at the
      compiler level. This works by passing the appropriate target architecture flag, such as ``-march=haswell`` for AVX2, to the compiler.  Unlike runtime dispatch, compile-time SIMD optimizations
```

https://github.com/apache/arrow/blob/3d467ac7bfae03cf2db09807054c5672e1959aec/cpp/cmake_modules/SetupCxxFlags.cmake#L47-L67

Author: I like this change, thanks.

>       cannot be changed at runtime (for example, if you compile Arrow C++ with
>       AVX512 enabled, the resulting binary will only run on AVX512-enabled
>       CPUs).  Setting ``ARROW_USER_SIMD_LEVEL=NONE`` prevents the execution of

Reviewer 4: ```suggestion
      cannot be changed at runtime. For example, if you compile Arrow C++ with
      AVX512 enabled, the resulting binary will only run on AVX512-enabled
      CPUs.  Setting ``ARROW_USER_SIMD_LEVEL=NONE`` prevents the execution of
```

   platforms currently do not implement any form of runtime dispatch.

   .. note::

>       In addition to runtime dispatch, Arrow's build system supports two

Reviewer 2: Ok, perhaps instead of making this note about CMake configuration variables, we should explain the two different kinds of SIMD optimizations? This would be more user-oriented (this is a page about environment variables after all).

Something like:
> In addition to runtime-selected SIMD optimizations dispatch, Arrow C++ can also be compiled
> with SIMD optimizations that are cannot be disabled at runtime. For example, by default, SSE4.2
> optimizations are enabled on x86 builds: therefore, with this default setting, Arrow C++ does
> not work at all on a CPU without support for SSE4.2. This setting can be changed using
> the``ARROW_SIMD_LEVEL`` CMake variable so as to either raise or lower the optimization level.
>
> Finally, the ``ARROW_RUNTIME_SIMD_LEVEL`` CMake variable sets a compile-time upper bound
> to runtime-selected SIMD optimizations. This useful in cases where a compiler reports support
> for an instruction set but does not actually support it in full.

Author: You make a good point about the approach. I like what you wrote and incorporated it in 46bcac8bec4896cec739ebb1151e82f1e54362f2. It's also shorter which is a plus.

      compile-time CMake variables for controlling SIMD: ``ARROW_SIMD_LEVEL``
      and ``ARROW_RUNTIME_SIMD_LEVEL``.

      The ``ARROW_SIMD_LEVEL`` variable sets the minimum supported SIMD level at

>       the compiler level. This works by passing the appropriate target

Reviewer 2: This is a bit misleading IMHO. It sets which instructions the compiler is allowed to emit regardless of dynamic dispatch.

Author: Ah, I think I see now. I just took a closer look at how ARROW_SIMD_LEVEL gets handled and the options don't stack onto each other. Borrowing from your language, how's this?

```suggestion
      The ``ARROW_SIMD_LEVEL`` variable controls which SIMD instructions the compiler is allowed to emit. 
 This works by passing the appropriate target
```

Reviewer 2: Hmm, this is still misleading in a different way. SIMD instructions are also emitted for runtime-selected optimizaitions, and they are not controlled by `ARROW_SIMD_LEVEL` :-)

Perhaps my suggestion was a dead end afterwards, and we shouldn't try to phrase this in terms of "emit[ting] SIMD instructions"?

      architecture flag, such as ``-march=haswell`` for AVX2, to the compiler.
      Unlike runtime dispatch, compile-time SIMD optimizations cannot be changed
      at runtime (for example, if you compile Arrow C++ with AVX512 enabled, the
      resulting binary will only run on AVX512-enabled CPUs).  Setting
      ``ARROW_USER_SIMD_LEVEL=NONE`` prevents the execution of explicit SIMD
      optimization code, but it does not rule out the execution of
      compiler-generated SIMD instructions.  E.g., on x86, Arrow is built with

>       ``ARROW_SIMD_LEVEL=SSE4_2`` by default.  The compiler may generate SSE4.2

Reviewer: Hmmm 🤔 A bit misleading here, since we've some hand-written SSE4.2 code, and some compiler generated code using some auto-vectorization techniques?

Author: I'm not familiar with either of those bits of the codebase but I can take a look tomorrow. Can you drop some links if you have them handy? (Logging off for the night here)

Reviewer: https://github.com/apache/arrow/issues/40333#issuecomment-1976015427

I've pasted some in issue

Author: Thanks. It looks like, in those cases, the symbols get compiled in but, at least for the example in level_comparison.cc, the code looks like it runs through the dispatch machinery in dispatch.h so it still wouldn't be selected if `ARROW_USER_SIMD_LEVEL` is set in such a way that prevents it. All that code is new to me so I'd appreciate another set of eyes.

Author: No worries, and thank you for the help with the text here @mapleFU. It's appreciated.

      instructions from any C/C++ source code.  On legacy x86 platforms which do
      not support SSE4.2, an Arrow binary may fail with SIGILL (Illegal

>       Instruction).  In this case, one must rebuild Arrow and PyArrow from

Reviewer 2: PyArrow shouldn't really be mentioned here, these are the C++ docs.

Author: ```suggestion
      Instruction).  In this case, one must rebuild Arrow from
```

Author: Removed the mention. Thanks.
